### DISCUSSION TITLE
Control Rig - Creating a new control rig - Als Control Rig Input not showing up in Variable type

### DISCUSSION THREAD

**Discussion by DavidShawGL** [Sixze/ALS-Refactored#441] [General]
Hey,

I am trying to create a new control rig from scratch and have noticed that I cannot create a new variable RigInput and assign it the variable type of Als Control Rig Input.

It just doesnt show up in the list when in a control rig blueprint. However if I create standard blueprint I can see it there.
![image](https://github.com/Sixze/ALS-Refactored/assets/55969504/0cec7170-2595-4f81-b1d6-95fb8c45eefe)

I can copy it from the ALS Control Rig and paste in my own, but this feels likes its either user error or a bug on my part.

The reason I am trying to create one from scratch is that we are not using ik retargeting and have our own skeleton that is not using the same orientations as the UE mannequin skeleton.

Its the same issue of the foot rotating 90 degrees on 2 axis as noted in this discussion: https://github.com/Sixze/ALS-Refactored/discussions/285

**Comment by DavidShawGL** [comment]
I just wanted to add we are on UE 5.3.2.

I wonder if the ALS control rig was created in an earler version and something has changed for why we cant set the variable type

**Comment by MrWilsonWilson** [comment]
not sure how come, 
![image](https://github.com/Sixze/ALS-Refactored/assets/122884397/ed04fb92-eed8-45ca-8489-338f4b50b78f)
 show plugin content maybe ?

**Reply by DavidShawGL** [reply]
We have it as a project plugin.

I'm not a coder, but looking at the c++ it shows the riginput setup as BlueprintReadWritable

**Comment by jomath229** [comment]
Something is definitely happening here, I'm having the same issue. Adding the plugin to a empty project, the control rig by default has the variable correctly setup. If you try adding a new Variable of type "FAlsControlRigInput", it doesn't seem to be detected. Some of the ALS struct are detected and others aren't. This is one of the weirdest behaviour I've seen in a while. Happening on 5.3.1. Maybe there's a way to acces it using a different path that's unknown from us.

![image](https://github.com/Sixze/ALS-Refactored/assets/49660042/9b80bb5b-7404-4831-a115-7d8eb430e563)

**Comment by jomath229** [comment]
Well after some testing, it seems uint8 aren't valid to be used in struct from the control rig blueprint. The ALS struct showing up are the one that don't have uint8. It is probable the control rig was originally created from a different Unreal Version that didn't have this bug and the meta data somehow still compiles correctly. Switch them up to some boolean and it will show up. Don't know why, I'll open up a bug report on unreal side.

**Reply by DavidShawGL** [reply]
ahh, thanks for investigating, one our our coders is coming to help me this afternoon so I can pass on the info.

Cheers
Dave

**Reply by DavidShawGL** [reply]
oh and I was incorrect in our source engine, 5.3.1.

I also found an issue that has been fixed in 5.3.2 where in the spine rotation function, from the For Each of SpineRotatioBones, when you Get Transform it will always have a warning due to it not unpacking the data properly.

**Reply by DavidShawGL** [reply]
Also as you noted, we changed the  ```uint8 <<NAME>> : 1 {false};``` to ```bool <<NAME>> = false``` and could select ALS Control Rig Input as a varaible.

The programmer helping me said something about packed uint8's (bit above me)

**Reply by jomath229** [reply]
Yeah uint8 were causing the issue, I guess something broke in 5.3.1 affecting them but since the source Ctrl rig from the plugin has probably been built on a different version than 5.3.1, it still works but not correctly. Kinda annoying having to update to 5.3.2, takes almost an entire day to rebuild the engine on my end. I'll stick to modifying the uint8 to bools for now until I update the engine later on.

**Reply by DavidShawGL** [reply]
We are shifting to 5.3.2 soon as well so will hopefully fix some of the other issues for us too.

Dave

**Reply by jomath229** [reply]
fyi I'm running both source version and epic launcher version and I updated my launcher version to 5.3.2 after your comment and still no luck on the uint8, the struct is still invisible after trying the plugin cloned straight from github in an empty 5.3.2 new project. 

Might have to wait for an official fix for uint8 or switching the uint8 to booleans in the plugin's source code.

**Reply by DavidShawGL** [reply]
yeah, sorry I might not have been clear on what 5.3.2 fixed for me.

Get Transform was also failing in 5.3.1 in the SpineRotation function when i tried to recreate it from scratch in another blueprint. It would not resove the item as valid from the For Each.  Testing in 5.3.2 resolved this issue

I still had the uint8 as bools when I tested that.

Cheers,
Dave

**Comment by Sixze** [comment]
It looks like this issue still exists in UE 5.3.2, so I changed `uint8` back to `bool` here: e2a393695d9fa80050ab6afc646cf29408f5fb49.

**Reply by DavidShawGL** [reply]
Thank you
